SampCert/DifferentialPrivacy/Abstract.lean

```diff
@@ -10,110 +10,82 @@ import SampCert.Foundations.Basic
 /-!
 # Differential Privacy
 
-This file defines a notion of a differential private system.
+This file defines an abstract system of differentially private operators.
 -/
 
 noncomputable section
 
 open Classical Nat Int Real ENNReal
 
-def NonZeroNQ (nq : List T → SLang U) :=
+/--
+A mechanism is nonzero on its entire domain
+-/
+def NonZeroNQ (nq : List T → PMF U) :=
   ∀ l : List T, ∀ n : U, nq l n ≠ 0
 
-def NonTopSum (nq : List T → SLang U) :=
-  ∀ l : List T, ∑' n : U, nq l n ≠ ⊤
-
 namespace SLang
 
 abbrev Query (T U : Type) := List T → U
-abbrev Mechanism (T U : Type) := List T → SLang U
+
+abbrev Mechanism (T U : Type) := List T → PMF U
 
 /--
 Product of mechanisms.
-
-Note that the second mechanism does not depend on the output of the first; this is in currently
-in contrast to the notions of composition found in the DP literature.
 -/
-def privCompose (nq1 : Mechanism T U) (nq2 : Mechanism T V) (l : List T) : SLang (U × V) := do
+def privCompose (nq1 : Mechanism T U) (nq2 : Mechanism T V) (l : List T) : PMF (U × V) := do
   let A ← nq1 l
   let B ← nq2 l
   return (A,B)
 
 /--
 Mechanism obtained by applying a post-processing function to a mechanism.
 -/
-def privPostProcess (nq : Mechanism T U) (pp : U → V) (l : List T) : SLang V := do
+def privPostProcess (nq : Mechanism T U) (pp : U → V) (l : List T) : PMF V := do
   let A ← nq l
   return pp A
 
-
 /--
 Abstract definition of a differentially private systemm.
 -/
 class DPSystem (T : Type) where
   /--
-  Notion of differential privacy with a paramater (ε-DP, ε-zCDP, etc)
+  Differential privacy proposition, with one real paramater (ε-DP, ε-zCDP, etc)
   -/
   prop : Mechanism T Z → ℝ → Prop
   /--
-  Noise mechanism (eg. Laplace, Discrete Gaussian, etc)
-  Paramaterized by a query, sensitivity, and the numerator/denominator ofa (rational) security paramater.
+  A noise mechanism (eg. Laplace, Discrete Gaussian, etc)
+  Paramaterized by a query, sensitivity, and a (rational) security paramater.
   -/
-  noise : Query T ℤ → ℕ+ → ℕ+ → ℕ+ → Mechanism T ℤ
+  noise : Query T ℤ → (sensitivity : ℕ+) → (num : ℕ+) → (den : ℕ+) → Mechanism T ℤ
   /--
-  Adding noise to a query makes it differentially private.
+  Adding noise to a query makes it private.
   -/
   noise_prop : ∀ q : List T → ℤ, ∀ Δ εn εd : ℕ+, sensitivity q Δ → prop (noise q Δ εn εd) (εn / εd)
   /--
-  Notion of privacy composes by addition.
+  Privacy composes by addition.
   -/
   compose_prop : {U V : Type} → [MeasurableSpace U] → [Countable U] → [DiscreteMeasurableSpace U] → [Inhabited U] → [MeasurableSpace V] → [Countable V] → [DiscreteMeasurableSpace V] → [Inhabited V] → ∀ m₁ : Mechanism T U, ∀ m₂ : Mechanism T V, ∀ ε₁ ε₂ ε₃ ε₄ : ℕ+,
     prop m₁ (ε₁ / ε₂) → prop m₂ (ε₃ / ε₄) → prop (privCompose m₁ m₂) ((ε₁ / ε₂) + (ε₃ / ε₄))
   /--
-  Notion of privacy is invariant under post-processing.
+  Requirement for postcomposition to hold.
+  -/
+  postprocess_prop_f : {U V : Type} -> (f : U -> V) -> Prop
```

Comment:
That's a bit mysterious. Do we have this because of pure DP that still requires surjectivity?